src/types.rs

```diff
@@ -1086,6 +1088,129 @@ impl SimpleAsn1Writable for GeneralizedTime {
     }
 }
 
+/// Used for parsing and writing ASN.1 `GENERALIZED TIME` values accepting
+/// fractional seconds value.
+/// See https://github.com/alex/rust-asn1/issues/491 for discussion.
+#[derive(Debug, Clone, PartialEq, PartialOrd, Hash, Eq)]
+pub struct GeneralizedTimeFractional {
+    datetime: DateTime,
+    nanoseconds: Option<u32>, // Up to 1 ns precision (10^9)
+}
+
+impl GeneralizedTimeFractional {
+    pub fn new(dt: DateTime, nanoseconds: Option<u32>) -> ParseResult<GeneralizedTimeFractional> {
+        if let Some(val) = nanoseconds {
+            if val < 1 || val >= 1e9 as u32 {
+                return Err(ParseError::new(ParseErrorKind::InvalidValue));
+            }
+        }
+
+        Ok(GeneralizedTimeFractional {
+            datetime: dt,
+            nanoseconds,
+        })
+    }
+
+    pub fn as_datetime(&self) -> &DateTime {
+        &self.datetime
+    }
+
+    pub fn nanoseconds(&self) -> Option<u32> {
+        self.nanoseconds
+    }
+}
+
+fn read_fractional_time(data: &mut &[u8]) -> ParseResult<Option<u32>> {
+    // We cannot use read_byte here because it will advance the pointer
+    // However, we know that the is suffixed by 'Z' so reading into an empty
+    // data should lead to an error.
+    if data.first() == Some(&b'.') {
+        *data = &data[1..];
+
+        let mut fraction = 0u32;
+        // Read up to 9 digits
+        let mut digits = 0;
+        while digits < 9 {
+            match data.first() {
+                Some(b) => {
+                    if !b.is_ascii_digit() {
+                        if digits == 0 {
+                            // We must have at least one digit
+                            return Err(ParseError::new(ParseErrorKind::InvalidValue));
+                        }
+                        break;
+                    }
+
+                    *data = &data[1..];
+                    fraction = fraction * 10 + (b - b'0') as u32;
+                }
+                None => {
+                    return Err(ParseError::new(ParseErrorKind::InvalidValue));
+                }
+            }
+            digits += 1;
+        }
```

Comment:
Nitpick: I think you can write this in a more idiomatic fashion with `data.iter().take(9)` to get an iterator of up to 9 bytes.